doc/HLD_NewBtree.md

```diff
@@ -0,0 +1,85 @@
+
+Motr implements BTree at the backend in order to handle metadata related operations. It can be internal metadata such as cob list (used by ios and repair), allocation data (used by ad stob), balloc (used by ios), global dix indices (used by dix), meta index (used by cas) or external metadata by applications like s3server using catalogues (used by cas, dix). This B-tree code should support better performance and upgrade.
+
+## Requirements
+
+The scope of b-tree development is defined by the following requirements:
+
+- [r.btree.clean] develop b-tree (b+tree) implementation from scratch;
+- [r.btree.r2p0] b-tree implementation has to be available in the first release (ldrr2p0) and is fully integrated with the rest of the code-base;
+- [r.btree.future-proof] after the first release, online conversion to the later versions should be possible. Future versions of b-tree can add more features;
+- [r.btree.features] b-tree format and algorithms should be ready to support advanced features that will be necessary in the future;
+- [r.btree.performance] performance properties of the implementation (throughput, operations per second, concurrency levels, memory and io footprints) should be sufficient for fulfilment of the overall product performance requirements;
+- [r.btree.quality] the implementation should be of high quality. Quality should be measured and attained by comprehensive testing and documentation;
+- [r.btree.scale] the implementation should be sufficiently scalable to support the overall product scalability requirements.
+
+## System Design
+
+### Approach
+--------------------------
+The Btree code implements a flavor of B tree known as B+ tree. The structure of B+ tree gives it an advantage of improved performance and stability over a regular B-tree. The varied meta-data structures will will need keys and values which can be either fixed in size or variable in size. Considering this, the Btree code supports 3-types of Btree node formats. Also to keep the Keys and Values closer to their host Btree node, we embed the keys and values within the node instead of maintaining a pointer to the keys and values.
+
+### Node Types
+--------------------------
+Every Btree node has a header which stores important information such as Type of Btree node, Count of records, Key size, Value size, … 
+Below is the description of all the Node types which are supported by the Btree code.
+
+a. Fixed Key Fixed Value Size (FF node type)
+
+  The FF node type has fixed sized keys and values. Following trees use this type of node format: balloc group extents, balloc group desc, cob fileattr omg, cas meta, cas dead idx and cas ct idx.
+In case of leaf node, we start populating keys from the left side and the values from the right side of free node space. Since all the Keys are of same size and all the Values are also of same size (though the Keys and Values may not be of same size), we can calculate the address of a particular key or value based on its index.
```

Comment:
> In case of leaf node, we start populating keys from the left side and the values from the right side of free

we start popluating keys from left and values on right irrespective of internal and leaf node.

We can first describe node structure and then say In case of leaf node, value part will hold address of child node. This will avoid confusion while reading.